### PR TITLE
Respect "soft_fail" for core async sensors

### PR DESCRIPTION
**Pull request by Lee-W**
This is the following PR to https://github.com/apache/airflow/pull/33196 and intends to solve the issue that the soft_fail argument might not work on all sensors.

This pull request addresses the issue of the core async sensors potentially raising non-AirflowSkipException even when `soft_fail` is set to True. It achieves this by modifying the `DateTimeTrigger` to include the `soft_fail` argument.

-----

Not related to this PR: I noticed that we are currently using separate sensors for synchronous and asynchronous operations in our core sensors. Do you think it would be beneficial to merge them into a single sensor, like what the providers do?

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by hussein-awala — CHANGES_REQUESTED** [review]
Let's wait #33424 before deciding if we need to merge this one.

**Comment by Lee-W** [comment]
> Let's wait #33424 before deciding if we need to merge this one.

Hi, I just checked #33424. We might still need this one, as the exceptions are raised in the `execute` method before we enter deferring. (but the solution in #33424 can still save us time handling these kinds of stuff in `execute_complete` which is still great!)

File: airflow/triggers/temporal.py [hunk 1]
```diff
@@ -34,18 +35,29 @@ class DateTimeTrigger(BaseTrigger):
     The provided datetime MUST be in UTC.
     """
 
-    def __init__(self, moment: datetime.datetime):
+    def __init__(self, moment: datetime.datetime, soft_fail: bool = False):
```
Comment (uranusjr) [inline airflow/triggers/temporal.py]:
```suggestion
    def __init__(self, moment: datetime.datetime, *, soft_fail: bool = False) -> None:
```

File: airflow/triggers/temporal.py [hunk 2]
```diff
@@ -34,18 +35,29 @@ class DateTimeTrigger(BaseTrigger):
...
         super().__init__()
         if not isinstance(moment, datetime.datetime):
-            raise TypeError(f"Expected datetime.datetime type for moment. Got {type(moment)}")
+            exc = TypeError(f"Expected datetime.datetime type for moment. Got {type(moment)}")
+            if soft_fail:
+                raise AirflowSkipException("Skipping due to soft_fail is set to True.") from exc
+            raise exc
```
Comment (uranusjr) [inline airflow/triggers/temporal.py]: Do we need to use `from` here? I feel it’d be more concise to combine the two error messages into `AirflowSkipException` instead.

Comment (Lee-W) [reply airflow/triggers/temporal.py]: `from` is not necessary here. Either way works fine for me. Let me change to what you suggested. 🙂

Comment (uranusjr) [reply airflow/triggers/temporal.py]: Let’s raise the skip exception directly (and include the type error message in it)

Comment (Lee-W) [reply airflow/triggers/temporal.py]: Sure. just updated it!

File: airflow/triggers/temporal.py (hunk 1 above)
Comment (Lee-W) [reply airflow/triggers/temporal.py]: Sure. just updated it!

**Review by uranusjr — APPROVED** [review]

**Review by uranusjr — APPROVED** [review]

**Review by aritra24 — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

File: airflow/triggers/temporal.py [hunk 3]
```diff
@@ -34,18 +35,30 @@ class DateTimeTrigger(BaseTrigger):
...
+            message = f"Expected datetime.datetime type for moment. Got {type(moment)}"
+            if soft_fail:
+                raise AirflowSkipException(f"{message}. {skipping_message_postfix}")
+            raise TypeError(message)
         # Make sure it's in UTC
         elif moment.tzinfo is None:
-            raise ValueError("You cannot pass naive datetimes")
+            message = "You cannot pass naive datetimes"
+            if soft_fail:
+                raise AirflowSkipException(f"{message}. {skipping_message_postfix}")
```
Comment (hussein-awala) [inline airflow/triggers/temporal.py]: How could raising an `AirflowSkipException` in the trigger pass the task state to skipped? The trigger exception is not propagated to task resuming if I am not mistaken. WDYT?

**Review by hussein-awala — CHANGES_REQUESTED** [review]
> Hi, I just checked #33424. We might still need this one, as the exceptions are raised in the execute method before we enter deferring. (but the solution in https://github.com/apache/airflow/pull/33424 can still save us time handling these kinds of stuff in execute_complete which is still great!)

I agree that #33424 is not completed yet, but I don't think that this PR could fix the issue. The concept of soft fail is if the sensor fails, we raise the specific skip exception which is processed by the job runner to change the state to skipped instead of failed.

Could you add a test for the operator? (the added test is for the trigger and not the task)

File: airflow/triggers/temporal.py (hunk 3 above)
Comment (Lee-W) [reply airflow/triggers/temporal.py]: Take `DateTimeSensorAsync` as an example. The second line will raise the `AirflowSkipException` before we execute `self.defer`

```python
        self.defer(
            trigger=DateTimeTrigger(moment=target_dttm, soft_fail=self.soft_fail),
            method_name="execute_complete",
        )
```

Comment (hussein-awala) [reply airflow/triggers/temporal.py]: What about:
```python
         try:
           self.defer(
              trigger=DateTimeTrigger(moment=target_dttm, soft_fail=self.soft_fail),
              method_name="execute_complete",
          )
        except Exception as e:
          raise AirflowException(str(e))
```
or just changing `raise TypeError(message)` and `raise ValueError` to `AirflowException`.

In both cases the exception will be cached by BaseSensor, and a `AirflowSkipException` will be raised when soft_fail is True. I'm trying to avoid complicating the trigger code, and duplicating the soft fail logic.

cc: @uranusjr @potiuk WDYT?

Comment (Lee-W) [reply airflow/triggers/temporal.py]: Or what do you think about my first commit? https://github.com/apache/airflow/pull/33403/commits/b2f2662ae1a11ea928aad57acd2892c763c2db25 I was trying not to duplicate the check logic and thus moved it to the trigger

Comment (hussein-awala) [reply airflow/triggers/temporal.py]: for me it's much better

**Comment by hussein-awala** [comment]
Since it's a problem in the initialization of the trigger, and it is not a runtime issue, what about raising an exception in parsing time, to give the user a fast feedback about the wrong parameters, for example:
```diff
diff --git a/airflow/sensors/date_time.py b/airflow/sensors/date_time.py
index 1425028870..a036d09e01 100644
--- a/airflow/sensors/date_time.py
+++ b/airflow/sensors/date_time.py
@@ -69,6 +69,7 @@ class DateTimeSensor(BaseSensorOperator):
             raise TypeError(
                 f"Expected str or datetime.datetime type for target_time. Got {type(target_time)}"
             )
+        self.trigger = DateTimeTrigger(moment=timezone.parse(self.target_time))
 
     def poke(self, context: Context) -> bool:
         self.log.info("Checking if the time (%s) has come", self.target_time)
@@ -87,7 +88,7 @@ class DateTimeSensorAsync(DateTimeSensor):
 
     def execute(self, context: Context):
         self.defer(
-            trigger=DateTimeTrigger(moment=timezone.parse(self.target_time)),
+            trigger=self.trigger,
             method_name="execute_complete",
         )
```
This way the dag parsing will fail, and the user will fix the issue.

WDYT?

**Comment by Lee-W** [comment]
> Since it's a problem in the initialization of the trigger, and it is not a runtime issue, what about raising an exception in parsing time, to give the user a fast feedback about the wrong parameters, for example:
> 
> ```diff
> diff --git a/airflow/sensors/date_time.py b/airflow/sensors/date_time.py
> index 1425028870..a036d09e01 100644
> --- a/airflow/sensors/date_time.py
> +++ b/airflow/sensors/date_time.py
> @@ -69,6 +69,7 @@ class DateTimeSensor(BaseSensorOperator):
>              raise TypeError(
>                  f"Expected str or datetime.datetime type for target_time. Got {type(target_time)}"
>              )
> +        self.trigger = DateTimeTrigger(moment=timezone.parse(self.target_time))
>  
>      def poke(self, context: Context) -> bool:
>          self.log.info("Checking if the time (%s) has come", self.target_time)
> @@ -87,7 +88,7 @@ class DateTimeSensorAsync(DateTimeSensor):
>  
>      def execute(self, context: Context):
>          self.defer(
> -            trigger=DateTimeTrigger(moment=timezone.parse(self.target_time)),
> +            trigger=self.trigger,
>              method_name="execute_complete",
>          )
> ```
> 
> This way the dag parsing will fail, and the user will fix the issue.
> 
> WDYT?

This might not be doable for https://github.com/apache/airflow/blob/a6d803303bf71a84e9e59e94d9c088e3120bedb5/airflow/sensors/time_delta.py#L65-L66

File: airflow/sensors/date_time.py [hunk 4]
```diff
@@ -85,9 +85,13 @@ class DateTimeSensorAsync(DateTimeSensor):
...
     """
 
+    def __init__(self, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self.trigger = DateTimeTrigger(moment=timezone.parse(self.target_time))
+
     def execute(self, context: Context):
         self.defer(
-            trigger=DateTimeTrigger(moment=timezone.parse(self.target_time)),
+            trigger=self.trigger,
```
Comment (uranusjr) [inline airflow/sensors/date_time.py:94]: Why is this change needed? I don’t see `self.trigger` used anywhere else.

**Comment by Lee-W** [comment]
hi @hussein-awala, could you please check whether it works for you when you have time? Thanks!

**Review by hussein-awala — APPROVED** [review]
